### PR TITLE
Add possibility to ignore tags

### PR DESCRIPTION
**Pull request by Wosin**
This should add a possibility to ignore specified tags as proposed in https://github.com/zio/zio/issues/7232.

I've modified slightly code to avoid having 9 different cases. Will test tomorrow morning and mark ready for review if works correctly.

I've also modified slightly a reporting code since currently if none of the test fulfill the requirements passed by `-t`, `-tags` or `-ignore-tags` we will just print `()`, which isn't too informative.

### CODE REVIEW

File: test/shared/src/main/scala/zio/test/Spec.scala [hunk 1]
```diff
@@ -137,6 +137,15 @@ final case class Spec[-R, +E](caseValue: SpecCase[R, E, Spec[R, E]]) extends Spe
   final def filterTags(f: String => Boolean)(implicit trace: Trace): Option[Spec[R, E]] =
     filterAnnotations(TestAnnotation.tagged)(_.exists(f))
 
+  /**
+   * Returns a new spec with only those suites and tests except for the ones
+   * with tags satisfying the predicate. If all tests or suites have tags that
+   * satisfy the specified predicate then returns `Some` with an empty suite
+   * with the root label if this is a suite or `None` otherwise.
+   */
+  final def filterNotTags(f: String => Boolean)(implicit trace: Trace): Option[Spec[R, E]] =
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/Spec.scala:146]: Can we delete this and just do `filterTags(!p)` to keep the API slightly smaller?

File: test/shared/src/main/scala/zio/test/TestAnnotationMap.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ import scala.collection.immutable.Map
 /**
  * An annotation map keeps track of annotations of different types.
  */
-final class TestAnnotationMap private (private val map: Map[TestAnnotation[Any], AnyRef]) { self =>
+final class TestAnnotationMap(private val map: Map[TestAnnotation[Any], AnyRef]) { self =>
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/TestAnnotationMap.scala]: Why this change?

File: test/shared/src/main/scala/zio/test/Spec.scala (hunk 1 above)
Comment (Wosin) [reply test/shared/src/main/scala/zio/test/Spec.scala:146]: I think the problem with `filterTags` is that even if we do `filterTags(tag => tag != "smth")` it will also filter out all tests that are not tagged. `exists` require at least element to be `true`.  We would need to negate the `exists` itself I think.

File: test/shared/src/main/scala/zio/test/TestAnnotationMap.scala (hunk 2 above)
Comment (Wosin) [reply test/shared/src/main/scala/zio/test/TestAnnotationMap.scala]: Used that for some testing, not needed.

**Review by swoogles — APPROVED** [review]
Was able to test things locally and it works as advertised 🙂 
Thanks for the contribution!
I've also written some unit tests for both `tags` and `ignore-tags` that we can add as soon as this is merged in.

File: test/shared/src/main/scala/zio/test/Spec.scala (hunk 1 above)
Comment (adamgfraser) [reply test/shared/src/main/scala/zio/test/Spec.scala:146]: Makes sense. Thanks!

**Review by adamgfraser — APPROVED** [review]
@Wosin Thanks for your work on this!
